receiver/elasticapmreceiver/internal/mappers/intakeV2ToSemConv.go

```diff
@@ -0,0 +1,97 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This file contains all the mapping from IntakeV2 fields to OTel Semantic Convention
+
+package mappers // import "github.com/elastic/opentelemetry-collector-components/receiver/elasticapmreceiver/internal/mappers"
+
+import (
+	"github.com/elastic/apm-data/model/modelpb"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	semconv "go.opentelemetry.io/collector/semconv/v1.27.0"
+)
+
+// Translates resource attributes from the Elastic APM model to SemConv resource attributes
+func TranslateToOtelResourceAttributes(event *modelpb.APMEvent, attributes pcommon.Map) {
+	attributes.PutStr(semconv.AttributeServiceName, event.Service.Name)
+	attributes.PutStr(semconv.AttributeServiceVersion, event.Service.Version)
+	if event.Service.Language != nil {
+		attributes.PutStr(semconv.AttributeTelemetrySDKLanguage, event.Service.Language.Name)
+	}
+	attributes.PutStr(semconv.AttributeTelemetrySDKName, "ElasticAPM")
+	if event.Service.Environment != "" {
+		attributes.PutStr(semconv.AttributeDeploymentEnvironmentName, event.Service.Environment)
+	}
+}
+
+// Translates transaction attributes from the Elastic APM model to SemConv attributes
+func TranslateIntakeV2TransactionToOTelAttributes(event *modelpb.APMEvent, attributes pcommon.Map) {
+	setHttpAttributes(event, attributes)
+
+	if event.Span.Message != nil {
+		attributes.PutStr(semconv.AttributeMessagingDestinationName, event.Transaction.Message.QueueName)
+		attributes.PutStr(semconv.AttributeMessagingRabbitmqDestinationRoutingKey, event.Transaction.Message.RoutingKey)
+
+		// This may need to be unified, see AttributeMessagingSystem for spans
+		attributes.PutStr(semconv.AttributeMessagingSystem, event.Service.Framework.Name)
+	}
+}
+
+// Translates span attributes from the Elastic APM model to SemConv attributes
+func TranslateIntakeV2SpanToOTelAttributes(event *modelpb.APMEvent, attributes pcommon.Map) {
+	if event.Http != nil {
+
+		setHttpAttributes(event, attributes)
+
+		if event.Url != nil {
+			attributes.PutStr(semconv.AttributeURLFull, event.Url.Full)
```

Comment:
Not sure if any of this inner fields can be nil, if so we might need to add a guard function for all the setters.

Reply:
This is actually a good point - we need to guard these. For now I just added another check. This was already setting empty strings.

In the spec the required fields are listed: https://github.com/elastic/apm-server/tree/main/docs/spec/v2

If any of those are missing, apm server rejects those documents. But `url.full` isn't required.

This whole question makes me think... we should probably discuss how to handle invalid documents - like when a required field is missing.

In any case - this specific one is handled.